Add win32 events alternative to unix signals

ashie opened the pull request:
<!--
Thank you for contributing to Fluentd!
Your commits need to follow DCO: https://probot.github.io/apps/dco/
And please provide the following information to help us make the most of your pull request:
-->

**Which issue(s) this PR fixes**: 
None

**What this PR does / why we need it**: 
Since Windows doesn't have UNIX signals, features realized by them such as flushing buffers aren't implemented on Windows.
This PR adds triggers for them by using win32 events. Currently only supervisor process supports them.

In addition, this PR adds a new command `fluent-ctl`. It sends a simple command to a fluentd process.
The main purpose of this command is that providing a kill command alternative on Windows.
On UNIX, it just only send a UNIX signal to a specified process.

```
Usage: fluent-ctl COMMAND PID_OR_SIGNAME
    
Commands:
  shutdown
  restart
  flush
  reload
```

SIGNAME can be specified by fluentd's "-x" or "--signame" option.
Note that this command doesn't support sending events to windows service.
Probably we need to trigger these events via `sc` command for windows service.

This PR also fixes a bug that the graceful shutdown feature doesn't work correctly when Fluentd is launched as Windows service. On this case, worker processes were killed forcedly before receiving `GRACEFUL_STOP` command.

**Docs Changes**:
Add `fluent-ctl` to https://docs.fluentd.org/deployment/command-line-option

**Release Note**: 
Same as title.

cosmo0920 commented on bin/fluentd-windows-cmd:
```diff
@@ -0,0 +1,9 @@
+$LOAD_PATH.unshift(File.join(__dir__, 'lib'))
+require 'fluent/command/windows_cmd'
+
+Fluent::WindowsCmd.new.call
```
Should we raise an error when using non-Windows environment?

ashie replied:
I renamed the command to fluentdctl and support also UNIX environment, it just only sends a UNIX signal to a specified process though.

cosmo0920 commented on bin/fluentd-windows-cmd:
```diff
@@ -0,0 +1,9 @@
+#!/usr/bin/env ruby
+
+require 'win32/event'
```
On non-Windows environment, `win32/event` shouldn't be installed as a dependency and loaded.
Should we use `Fluent.windows?` for guarding to require on non-Windows environment?

cosmo0920 replied:
Or, just raise LoadError here is enough?

ashie replied:
Ditto

ashie commented:
Matters to be discussed:

* I'm not sure the command name `fluentdctl` is appropriate or not.
  * `fluentd-ctl`? `fluentctl`? `fluent-ctl`?
* Although I've simplified the command names ASAP, I'm not sure they are appropriate or not.
  * e.g.) omit `buffers` of `flush` command, omit `graceful` of `reload` command
* UNIX version of `fluentdctl` may not be needed.
  * It just only send a UNIX signal to a specified process.

cosmo0920 commented:
I'd confirmed that `fluentdctl` can work on both of Windows and Linux.

> * I'm not sure the command name `fluentdctl` is appropriate or not.
>    * `fluentd-ctl`? `fluentctl`? `fluent-ctl`?

For current command naming convention, I vote `fluentctl` or `fluent-ctl`.

ashie commented:
> For current command naming convention, I vote `fluentctl` or `fluent-ctl`.

I've renamed it to fluent-ctl.

ashie commented:
Now I've found that both the original implementation for shutdown and this PR doesn't support multiple workers #3066.

ashie commented:
Now I''ve found that fluent-ctl can't send events when fluentd is run as windows service, even if fluent-ctl is run as "SYSTEM" user.
So that I renamed the 3rd argument of the command from `PID_OR_SVCNAME` to `PID_OR_SIGNAME`, and removed the default parameter for it.

For Windows service, probably we need to trigger the events via `sc` command.

ashie commented:
> Now I've found that both the original implementation for shutdown and this PR don't support multiple workers: #3066

It's ServerEngine's fault, I'll fix it. This PR should follow the change.

ashie commented:
> Now I've found that fluent-ctl can't find events when fluentd is run as windows service, even if fluent-ctl is run as "SYSTEM" user.

Windows Service processes are separated as "Session 0" as of Windows Vista.
[PsExec](https://docs.microsoft.com/en-us/sysinternals/downloads/psexec) has a option to execute a command on "Session 0".
Now I've confirmed that I can trigger the events by the following command
```
\path\to\psexec.exe -s -i 0 \path\to\fluent-ctl restart fluentdwinsvc
```

ashie commented:
> This PR also fixes a bug that the graceful shutdown feature doesn't work correctly when Fluentd is launched as Windows service.

I've filed a new pull request to fix it: #3156

ashie commented:
The remaining issue of this PR is:

> SIGNAME can be specified by fluentd's "-x" or "--signame" option.
> Note that this command doesn't support sending events to windows service.
> Probably we need to trigger these events via sc command for windows service.

I'm now working on it.
`sc` command can send pre-defined control codes or user-defined control codes to a service:

https://docs.microsoft.com/en-us/windows/win32/api/winsvc/nf-winsvc-controlservice

* For `USER2` signal (gaceful-reload feature) on UNIX, probably mapping it to `SERVICE_CONTROL_PARAMCHANGE`  is suitable.
* For `USER1` (flush) and `HUP` (restart), there is no suitable pre-defined control code, so I'll map these features to user-defined control codes.
  * But [win32-sercvice](https://github.com/chef/win32-service/blob/master/lib/win32/daemon.rb) doesn't support receiving user-defined control codes. I'll send PR to it to enable receiving them.
  * Or we can map these features to rarely used control codes such as `SERVICE_CONTROL_NETBIND*` (deprecated control codes), but it's not so good design.

ashie commented:
>  * But [win32-sercvice](https://github.com/chef/win32-service/blob/master/lib/win32/daemon.rb) doesn't support receiving user-defined control codes. I'll send PR to it to enable receiving them.

https://github.com/chef/win32-service/pull/74

ashie commented:
And I've pushed the patch to send /receive a control code to/from fluend windows service.

shutdown:
```
# sc control 1 fluentdwinsvc
or
# sc control stop fluentdwinsvc
or
# fluent-ctl shutdown
```

reload:
```
# sc control paramchange fluentdwinsvc
or
# sc control  6 fluentdwinsvc
or
# fluent-ctl reload 
```

restart (need the patch for win32-service):
```
# sc control fluentdwinsvc 128
or
# fluent-ctl restart
```

flush (need the patch for win32-service):
```
# sc control fluentdwinsvc 129
or
# fluent-ctl flush
```

ashie commented:
I already completed this work but we need to wait next release of win32-service (probably 2.3.0).

repeatedly commented on lib/fluent/supervisor.rb:
```diff
@@ -159,37 +165,101 @@ def stop_counter_server
     end
 
+    def stop_windows_event_thread
+      if Fluent.windows?
```
L73 is `stop_windows_event_thread if Fluent.windows?`.
Is this `if` for more safety?

ashie replied:
> Is this if for more safety?

That's right.

repeatedly commented on lib/fluent/supervisor.rb:
```diff
@@ -745,33 +831,45 @@ def install_main_process_signal_handlers
-        command_pipe.binmode
-        command_pipe.sync = true
+    def install_main_process_command_handlers
+      command_pipe = $stdin.dup
```
Why change from `STDIN` to `$stdin`?

ashie replied:
For testing incoming commands at workers. See `test_main_process_command_handlers`.

ashie commented on lib/fluent/command/ctl.rb:
```diff
@@ -0,0 +1,179 @@
+        flush: 129,
+        reload: SERVICE_CONTROL_PARAMCHANGE,
+      }
+      COMMAND_MAP = WINDOWS_EVENT_MAP
```
No particular reason. I'll remove them.

cosmo0920 commented on test/test_supervisor.rb:
```diff
@@ -190,6 +248,38 @@ def test_rpc_server
     $log.out.reset if $log.out.is_a?(Fluent::Test::DummyLogDevice)
   end
 
+  def test_rpc_server_windows
+    omit "Only for windows platform" unless Fluent.windows?
+
+    create_info_dummy_logger
+
+    opts = Fluent::Supervisor.default_options
+    sv = Fluent::Supervisor.new(opts)
+    conf_data = <<-EOC
+  <system>
+    rpc_endpoint 0.0.0.0:24447
+  </system>
+    EOC
+    conf = Fluent::Config.parse(conf_data, "(test)", "(test_dir)", true)
+    sys_conf = sv.__send__(:build_system_config, conf)
+
+    server = DummyServer.new
+    def server.config
+      {
+        :signame => "TestFluentdEvent",
+        :worker_pid => 5963,
+      }
+    end
+    server.rpc_endpoint = sys_conf.rpc_endpoint
+
+    server.run_rpc_server
+
+    mock(server).restart(true) { nil }
+    Net::HTTP.get URI.parse('http://127.0.0.1:24447/api/plugins.flushBuffers')
+
+    server.stop_rpc_server
+  end
```
`Net::HTTP.get URI.parse('http://127.0.0.1:24447/api/plugins.flushBuffers')`'s return value should be checked with some assertions?

cosmo0920 commented:
win32-service 2.3.0 which includes user-defined control code handling has been released:
https://github.com/chef/win32-service/commit/bde5569a9512f2f7ae07a754ea29f4c7b7b1abd0

cosmo0920 commented:
We should bump up dependent win32-service gem version to 2.3.0 before merging into master:
https://github.com/fluent/fluentd/blob/master/fluentd.gemspec#L36

cosmo0920 reviewed (approved):
I also confirmed that `fluent-ctl` can work for `shutdown/restart/flush/reload` on Windows Service.

BTW, I feel that fluent-ctl start is missing for Windows Service....
Starting Windows Service should be executed from sc/net command or Windows Management tool's Service?

ashie commented:
> Starting Windows Service should be executed from sc/net command or Windows Management tool's Service?

That's right.

repeatedly commented:
If there is no more comments, I will merge this patch.

cosmo0920 commented:
> If there is no more comments, I will merge this patch.

:+1:

Bumping up dependent win32-service gem version to v2.3.0 should do later.
win32-service v2.3.0 is not released yet.

cosmo0920 commented:
Let's :shipit:

cosmo0920 commented:
:tada:

chook commented:
this is amazing, what about SIGCONT or forcing a thread dump @ashie
